node/core/prospective-parachains/src/tests.rs

```diff
@@ -0,0 +1,1332 @@
+		back_candidate(&mut virtual_overseer, &candidate_b, candidate_hash_b).await;
+
+		// Get backable candidate.
+		get_backable_candidate(
+			&mut virtual_overseer,
+			&leaf_a,
+			1.into(),
+			vec![],
+			Some(candidate_hash_a),
+		)
+		.await;
+		get_backable_candidate(
+			&mut virtual_overseer,
+			&leaf_a,
+			1.into(),
+			vec![candidate_hash_a],
+			Some(candidate_hash_b),
+		)
+		.await;
+
+		// Should not get anything at the wrong path.
+		get_backable_candidate(
+			&mut virtual_overseer,
+			&leaf_a,
+			1.into(),
+			vec![candidate_hash_b],
+			None,
+		)
+		.await;
+
+		virtual_overseer
+	});
+
+	assert_eq!(view.active_leaves.len(), 1);
+	assert_eq!(view.candidate_storage.len(), 2);
+	// Two parents and two candidates on para 1.
+	assert_eq!(view.candidate_storage.get(&1.into()).unwrap().len(), (2, 2));
+	assert_eq!(view.candidate_storage.get(&2.into()).unwrap().len(), (0, 0));
+}
+
+// Test depth query.
+#[test]
+fn check_depth_query() {
+	let test_state = TestState::default();
+	let view = test_harness(|mut virtual_overseer| async move {
+		// Leaf A
+		let leaf_a = TestLeaf {
+			number: 100,
+			hash: Hash::from_low_u64_be(130),
+			para_data: vec![
+				(1.into(), PerParaData::new(97, HeadData(vec![1, 2, 3]))),
+				(2.into(), PerParaData::new(100, HeadData(vec![2, 3, 4]))),
+			],
+		};
+
+		// Activate leaves.
+		activate_leaf(&mut virtual_overseer, &leaf_a, &test_state).await;
+
+		// Candidate A.
+		let (candidate_a, pvd_a) = make_candidate(
+			&leaf_a,
+			1.into(),
+			HeadData(vec![1, 2, 3]),
+			HeadData(vec![1]),
+			test_state.validation_code_hash,
+		);
+		let candidate_hash_a = candidate_a.hash();
+		let response_a = vec![(leaf_a.hash, vec![0])];
+
+		// Candidate B.
+		let (candidate_b, pvd_b) = make_candidate(
+			&leaf_a,
+			1.into(),
+			HeadData(vec![1]),
+			HeadData(vec![2]),
+			test_state.validation_code_hash,
+		);
+		let candidate_hash_b = candidate_b.hash();
+		let response_b = vec![(leaf_a.hash, vec![1])];
+
+		// Candidate C.
+		let (candidate_c, pvd_c) = make_candidate(
+			&leaf_a,
+			1.into(),
+			HeadData(vec![2]),
+			HeadData(vec![3]),
+			test_state.validation_code_hash,
+		);
+		let candidate_hash_c = candidate_c.hash();
+		let response_c = vec![(leaf_a.hash, vec![2])];
+
+		// Get hypothetical depth of candidate A before adding it.
+		get_hypothetical_depth(
+			&mut virtual_overseer,
+			candidate_hash_a,
+			1.into(),
+			leaf_a.para_data(1.into()).head_data.clone(),
+			leaf_a.hash,
+			leaf_a.hash,
+			vec![0],
+		)
+		.await;
+
+		// Add candidate A.
+		second_candidate(
+			&mut virtual_overseer,
+			candidate_a.clone(),
+			pvd_a.clone(),
+			response_a.clone(),
+		)
+		.await;
+
+		// Get depth of candidate A after adding it.
+		get_hypothetical_depth(
+			&mut virtual_overseer,
+			candidate_hash_a,
+			1.into(),
+			HeadData(vec![1, 2, 3]),
+			// TODO: When would these be different?
+			leaf_a.hash,
+			leaf_a.hash,
```

Comment:
Fragment tree relay parent is a leaf, while candidate's relay parent may be in allowed ancestry.